### PR TITLE
This PR squashes a bunch of warnings...

### PR DESCRIPTION
**Pull request by rhornung67**
...due to unused variables, and the facts that C++14 will not implictly declare 'const' non-static member functions that are marked constexpr and class member functions marked 'default' should not be marked with '__host__ __device__' because the host and device compilers will not generally emit the same simple default implementations.

Due to inconsistencies between newer and older versions of nvcc, we will live with the host/device default members until we have a better solution.  Nevertheless, this PR does squash a lot of warnings.

### CODE REVIEW

File: include/RAJA/pattern/detail/reduce.hpp [hunk 1]
```diff
@@ -120,7 +120,7 @@ class ValueLoc
   constexpr ValueLoc() = default;
   constexpr ValueLoc(ValueLoc const &) = default;
 
-  RAJA_HOST_DEVICE ValueLoc &operator=(ValueLoc const &) = default;
+  ValueLoc& operator=(ValueLoc const &) = default;
```
Comment (davidbeckingsale) [inline include/RAJA/pattern/detail/reduce.hpp]: I think this change has broken stuff (see Travis).

File: include/RAJA/pattern/detail/reduce.hpp [hunk 2]
```diff
@@ -120,7 +120,7 @@ class ValueLoc
   constexpr ValueLoc() = default;
   constexpr ValueLoc(ValueLoc const &) = default;
 
-  RAJA_HOST_DEVICE ValueLoc &operator=(ValueLoc const &) = default;
```
Comment (rchen20) [inline include/RAJA/pattern/detail/reduce.hpp]: RAJA_HOST_DEVICE here is necessary for SoAArray/Ptr in the CUDA reducer implementation. We probably need to leave this in and live with the warnings. Looping in @MrBurmark .

File: include/RAJA/pattern/detail/reduce.hpp (hunk 1 above)
Comment (rhornung67) [reply include/RAJA/pattern/detail/reduce.hpp]: Yeah, it removes warnings with newer nvcc versions, but breaks older ones (like what we use on travis). I'll revert this.

File: include/RAJA/pattern/detail/reduce.hpp (hunk 2 above)
Comment (rhornung67) [reply include/RAJA/pattern/detail/reduce.hpp]: Yes, I will revert this. See my comment above.

Comment (davidbeckingsale) [reply include/RAJA/pattern/detail/reduce.hpp]: Sounds good. We need to decide on CUDA version compatibility. If we are going to move up to only supporting 10+, I can update the Travis images.

Comment (rhornung67) [reply include/RAJA/pattern/detail/reduce.hpp]: Well, we need to discuss this. I've been testing some things on Argonne machines for ATPESC and they have really old compilers.

**Comment by trws** [comment]
We can squash this one for both if we add explicit implementation with
the annotation, or do an ifdef.

**Comment by rhornung67** [comment]
@trws working on it now.

**Pull request merged**